File: Bin/TerrainEditFilters/riverbuilder.lua
-- New road builder

return
{
	name="River Builder v1.0",
	description="Construct a river from the current list of waypoints.",
	options=
	{
		{name="Starting Bed width", type="value", value=2},
		{name="Ending Bed Width", type="value", value=64},
		{name="Bed Hardness", type="value", value=0.125},
		{name="Paving Width", type="value", value=70},
		{name="Paving Hardness", type="value", value=0.125},
		{name="Paving Layer", type="value", value=2},
		{name="Segment steps", type="value", value=10},
		{name="Start Depth", type="value", value=0.1},
		{name="End Depth", type="value", value=0.2},
	},
	
	execute=function(self)
		local startbedwidth=self.options[1].value
		local endbedwidth=self.options[2].value
		local bedhardness=self.options[3].value
		local pavingwidth=self.options[4].value
		local pavinghardness=self.options[5].value
		local pavinglayer=self.options[6].value
		local segments=self.options[7].value
		local startdepth=self.options[8].value
		local enddepth=self.options[9].value
		
		local buffer=RasterBuffer(hmap:GetWidth(), hmap:GetHeight())
		local blend=RasterBuffer(hmap:GetWidth(), hmap:GetHeight())
		blend:fill(0)
		
		local c
		
		local plist=RasterVertexList()
		for _,c in ipairs(waypoints) do
			local pos=c.position
			local norm=WorldToNormalized(hmap,terrain,pos)
			local hx=math.floor(norm.x*hmap:GetWidth())
			local hy=math.floor(norm.y*hmap:GetHeight())
			local ht=GetHeightValue(hmap,hx,(hmap:GetHeight()-1)-hy)
			plist:push_back(RasterVertex(hx,hy,ht))
		end
		
		local curve=RasterVertexList()
		TessellateLineList(plist, curve, segments)
		
		-- Iterate the points from start to finish and set height
		local lastht=curve:at(0).val_
		for c=0,curve:size()-1,1 do
			local v=curve:at(c)
			v.val_=math.min(lastht, v.val_)
			lastht=v.val_
		end
		
		for c=0,curve:size()-1,1 do
			local v=curve:at(c)
			local t=c/(curve:size()-1)
			local deep=startdepth+t*(enddepth-startdepth)
			v.val_=math.max(0, v.val_-deep)
		end
		
		-- Build road bed quadlist and rasterize
		local quad=RasterVertexList()
		BuildQuadStripVarying(curve, quad, startbedwidth, endbedwidth)
		RasterizeQuadStrip(buffer, quad)
		
		-- Iterate roadbed quadlist and set alternate verts to 0 and 1, rasterize to blend and apply the paving modifier
		for c=0,quad:size()-1,1 do
			local v=quad:at(c)
			v.val_=c % 2
		end
		RasterizeQuadStrip(blend, quad)
		ApplyBedFunction(blend, bedhardness, true)
		
		BlendHeightWithRasterizedBuffer(hmap, buffer, blend)
		terrain:ApplyHeightMap()
		
		quad=RasterVertexList()
		BuildQuadStripVarying(curve, quad, startbedwidth, endbedwidth)
		blend:fill(0)
		local color
		if pavinglayer==0 then color=Color(1,0,0,0)
		elseif pavinglayer==1 then color=Color(0,1,0,0)
		elseif pavinglayer==2 then color=Color(0,0,1,0)
		else color=Color(0,0,0,1)
		end
		
		for c=0,quad:size()-1,1 do
			local v=quad:at(c)
			v.val_=c % 2
		end
		RasterizeQuadStrip(blend, quad)
		ApplyBedFunction(blend, pavinghardness, true)
		--BlendColorWithRasterizedBuffer(blend1, blend, color)
		BlendRasterizedBuffer8(blend1,blend2,blend,pavinglayer,mask,usemask,invertmask)
		blendtex1:SetData(blend1,false)
		blendtex2:SetData(blend2,false)
	end,
}

File: Source/Filters/riverbuilder.cpp
#include "riverbuilder.h"
#include "../rasterization.h"

RiverBuilderFilter::RiverBuilderFilter(Context *context,TerrainContext *tc, WaypointGroupUI *wg) : FilterBase(context, tc, wg)
{
	name_="River Builder";
	description_="Construct a river bed along a spline strip.";
	options_=
	{
		{"Spline: ", (unsigned int)OT_SPLINE},
		{"Starting bed width: ", 32.0f},
		{"Ending bed width: ", 64.f},
		{"Starting depth: ", 0.1f},
		{"Ending depth: ", 0.2f},
		{"Bed hardness: ", 0.5f},
		{"Paving hardness: ", 0.5f},
		{"Segment steps: ", 10.f},
		{"Paving layer: ", {"Layer 0", "Layer 1", "Layer 2", "Layer 3", "Layer 4", "Layer 5", "Layer 6", "Layer 7"}},
		{"Use Mask 0? ", false},
		{"Invert Mask 0? ", false},
		{"Use Mask 1? ", false},
		{"Invert Mask 1? ", false},
		{"Use Mask 2? ", false},
		{"Invert Mask 2? ", false}
	};
}

void RiverBuilderFilter::Execute()
{
	float startingbedwidth=options_[1].value_;
	float endingbedwidth=options_[2].value_;
	float startingdepth=options_[3].value_;
	float endingdepth=options_[4].value_;
	float bedhardness=options_[5].value_;
	float pavinghardness=options_[6].value_;
	int segments=(int)options_[7].value_;

	const ea::string &sel=options_[8].listSelection_;
	unsigned int which=0;
	if(sel=="Layer 1") which=1;
	else if(sel=="Layer 2") which=2;
	else if(sel=="Layer 3") which=3;
	else if(sel=="Layer 4") which=4;
	else if(sel=="Layer 5") which=5;
	else if(sel=="Layer 6") which=6;
	else if(sel=="Layer 7") which=7;

	MaskSettings ms(options_[9].flag_, options_[10].flag_, options_[11].flag_, options_[12].flag_, options_[13].flag_, options_[14].flag_);

	IntVector2 tsize=terrainContext_->GetTerrainMapSize();
	CArray2Dd buffer(tsize.x_, tsize.y_);
	CArray2Dd blend(tsize.x_, tsize.y_);
	blend.fill(0);

	std::vector<Vector3> &knots=options_[0].splineKnots_;
	RasterVertexList plist;
	for(auto k : knots)
	{
		Vector2 norm=terrainContext_->WorldToNormalized(k);
		int hx=std::floor(norm.x_ * (float)tsize.x_);
		int hy=std::floor(norm.y_ * (float)tsize.y_);
		float ht=terrainContext_->GetHeightValue(hx, (tsize.y_-1)-hy);
		plist.push_back(RasterVertex(hx, hy, ht));
	}

	RasterVertexList curve;
	TessellateLineList(&plist, &curve, segments);

	if(curve.size()==0) return;
	float lastht=curve[0].val_;
	for(unsigned int c=0; c<curve.size(); ++c)
	{
		auto &v=curve[c];
		v.val_=std::min(lastht, v.val_);
		lastht=v.val_;
	}

	for(unsigned int c=0; c<curve.size(); ++c)
	{
		auto &v=curve[c];
		float t=(float)c/(float)(curve.size()-1);
		float deep=startingdepth+t*(endingdepth-startingdepth);
		v.val_=std::max(0.0f, v.val_-deep);
	}

	RasterVertexList quad;
	BuildQuadStripVarying(&curve, &quad, startingbedwidth, endingbedwidth);
	RasterizeQuadStrip(&buffer, &quad);

	for(unsigned int c=0; c<quad.size(); ++c)
	{
		RasterVertex &v=quad[c];
		v.val_=c % 2;
	}
	RasterizeQuadStrip(&blend, &quad);
	ApplyBedFunction(&blend, bedhardness, true);
	terrainContext_->BlendHeightBuffer(buffer, blend, ms);

	quad.clear();
	BuildQuadStripVarying(&curve, &quad, startingbedwidth, endingbedwidth);
	blend.fill(0);

	for(unsigned int c=0; c<quad.size(); ++c)
	{
		RasterVertex &v=quad[c];
		v.val_=c%2;
	}
	RasterizeQuadStrip(&blend, &quad);
	ApplyBedFunction(&blend, pavinghardness, true);
	terrainContext_->SetLayerBufferMax(blend, which, ms);
}
